File: tools/rvp.py
# $Id$

import sys, os, fcntl, string, re, xml.parsers.expat

global rvpin,rvpout,pat,errors,parser,text,mixed

class ProtocolError(Exception):
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)
 
def launch():
	global rvpin,rvpout
	rvpin,rvpout=os.popen2('rvp '+string.join(sys.argv[1:],' '),0)

def send(s):
	rvpin.write(s.encode('UTF-8')+'\0')
	rvpin.flush()
	
def recv():
	s=''
	while 1:
		c=rvpout.read(1)
		if(c=='\0'): break
		s=s+c
	return s

def resp():
	global errors
	r=string.split(recv(),' ',3)
	if(r[0]=='ok'):
		return r[1]
	if(r[0]=='error'):
		errors=1
		if(r[3]!=''):
			sys.stderr.write(str(parser.lineno)+","+str(parser.offset)+": "+r[3])
		return r[1]
	raise ProtocolError,"unexpected response '"+r[0]+"'"

def start_tag_open(cur,name):
	send('start-tag-open '+cur+' '+name)
	return resp()

def attribute(cur,name,val):
	send('attribute '+cur+' '+name+' '+val)
	return resp()
	
def start_tag_close(cur,name):
	send('start-tag-close '+cur+' '+name)
	return resp()

def end_tag(cur,name):
	send('end-tag '+cur+' '+name)
	return resp()

def textonly(cur,text):
	send('text '+cur+' '+text)
	return resp()

# in mixed content, whitespace is simply discarded, and any
# non-whitespace is equal; but this optimization gives only only
# 5% increase in speed at most in practical cases
def mixed(cur,text):
	if(re.search('[^\t\n ]',text)):
		send('mixed '+cur+' .')
		return resp()
	else:
		return cur

def start(g):
	send('start '+g)
	return resp()

def quit():
	send('quit')
	return resp()

# Expat handlers
def flush_text():
	global mixed,pat,text
	
	if(pat==mixed):
		mixed(pat,text)
	else:
		textonly(pat,text)
	text=''

def start_element(name,attrs):
	global mixed,pat

	mixed=1
	flush_text()
	pat=start_tag_open(pat,name)
	mixed=0
	for n,v in attrs.items():
		pat=attribute(pat,n,v)
	pat=start_tag_close(pat,name)

def end_element(name):
	global mixed,pat

	flush_text()
	pat=end_tag(pat,name)
	mixed=1

def characters(data):
	global text
	
	text=text+data

# Main

errors=0
launch()
pat=start('0')

parser=xml.parsers.expat.ParserCreate('UTF-8',':')
parser.StartElementHandler=start_element
parser.EndElementHandler=end_element
parser.CharacterDataHandler=characters

text=''
parser.ParseFile(sys.stdin)

quit()

File: tools/rvp.pl
#!/usr/local/bin/perl
# $Id$

# embedding sample for RVP, a part of RNV, http://davidashen.net/rnv.html
# code kept simple to show the technique, not to provide a general purpose
# module.
#
# details of the protocol are in a long comment near the start of rvp.c
#

use FileHandle; use IPC::Open2;
use XML::Parser::Expat;

use strict;

my @RVP=("rvp","-s");

my ($parser,$errors); # declared here for use in resp()

# rvp wrapper
$|=1;  # using pipes
$/="\0"; # the zero byte is separator

# write queries to RVPIN, get responses from RVPOUT
open2(\*RVPOUT,\*RVPIN,@RVP,shift @ARGV);

sub resp {
  $_=<RVPOUT>;
  chop;
  /^ok (\d+).*/ and return $1;
  /^error (\d+) (\d+) (.*)/ and do {
      my ($pat,$msg)=($1,$3);

     # if the message is empty, don't print the message, 
     # the error has occured in already erroneous state 
      $msg and printf STDERR "%u,%u: %s\n",
        $parser->current_line(),
        $parser->current_column(),
        $msg;
      $errors=1;
      return $pat;
    };
  die "protocol error, stopped ";
}

sub start_tag_open {
  my($cur,$name)=@_;
  print RVPIN "start-tag-open $cur $name\0";
  return resp();
}

sub attribute {
  my($cur,$name,$val)=@_;
  print RVPIN "attribute $cur $name $val\0";
  return resp();
}

sub start_tag_close {
  my($cur,$name)=@_;
  print RVPIN "start-tag-close $cur $name\0";
  return resp();
}

sub end_tag {
  my($cur,$name)=@_;
  print RVPIN "end-tag $cur $name\0";
  return resp();
}

sub text {
  my($cur,$text)=@_;
  print RVPIN "text $cur $text\0";
  return resp();
}

sub mixed {
  my($cur,$text)=@_;
  print RVPIN "mixed $cur $text\0";
  return resp();
}

sub start {
  print RVPIN "start\0";
  return resp();
}

sub quit {
  print RVPIN "quit\0";
  return resp();
}

my ($current,$text,$mixed)=(0,"",0);

# Expat handlers

# Expat does not merge cdata into one text node;
# application has to do it explicitely (see characters())
sub flush_text {
  $current=$mixed?mixed($current,$text):text($current,$text);
  $text="";
}

# last colon in a name separates the local name from the URI
sub qname {
   my ($p,$name)=@_;
   return join(':',$p->namespace($name),$name);
}
   
sub start_element {
  my ($p,$el,%attrs)=@_;

  $mixed=1;
  flush_text();
  $current=start_tag_open($current,qname($p,$el));
  $mixed=0;
  for my $atname (keys %attrs) {
    $current=attribute($current,qname($p,$atname),$attrs{$atname});
  }
  $current=start_tag_close($current,qname($p,$el));
}

sub end_element {
  my ($p,$el)=@_;
  flush_text();
  $current=end_tag($current,qname($p,$el));
  $mixed=1;
}

sub characters {
  my ($p,$chars)=@_;
  $text=$text.$chars;
}

# Main

$errors=0;
$parser=new XML::Parser::Expat(
  Namespaces=>1);

$parser->setHandlers(
  Start=>\&start_element,
  End=>\&end_element,
  Char=>\&characters);

$current=start();
$parser->parse(*STDIN);
quit();

exit($errors);
